src/tiered/tiered_handle.c

```diff
@@ -791,13 +791,42 @@ __wt_tiered_open(WT_SESSION_IMPL *session, const char *cfg[])
     return (ret);
 }
 
+/*
+ * __tiered_cleanup --
+ *     Cleanup a tiered data handle.
+ */
+static void
+__tiered_cleanup(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+{
+    uint32_t i;
+
+    /* Cleanup and dereference all tiers.  */
+    for (i = 0; i < WT_TIERED_MAX_TIERS; i++) {
+        if (tiered->tiers[i].tier != NULL && tiered->tiers[i].tier->session_inuse > 0)
+            (void)__wt_atomic_subi32(&tiered->tiers[i].tier->session_inuse, 1);
+        tiered->tiers[i].tier = NULL;
+        tiered->tiers[i].flags = 0;
+        __wt_free(session, tiered->tiers[i].name);
+    }
+
+    /* Cleanup other fields in tiered structure. */
+    __wt_free(session, tiered->key_format);
+    __wt_free(session, tiered->value_format);
+    __wt_free(session, tiered->obj_config);
+    tiered->current_id = tiered->next_id = tiered->oldest_id = 0;
+    tiered->flags = 0;
+    tiered->bstorage = NULL;
+}
+
 /*
  * __wt_tiered_close --
  *     Close a tiered data handle.
  */
 int
-__wt_tiered_close(WT_SESSION_IMPL *session)
+__wt_tiered_close(WT_SESSION_IMPL *session, WT_TIERED *tiered)
```

Comment:
I *think* this should not be necessary. `tiered_cleanup` should be able to get the tiered structure from `session->dhandle` like `tiered_open` and `tiered_switch` do. The `session->dhandle` should be set up on this call I believe.

Reply:
I agree with you here, and I was reluctant to make this change. But looking at  `__wt_tiered_discard`, `__wt_schema_close_table` and `__wt_tiered_tree_close` I thought I should make `__wt_tiered_close` look consistent in its "neighborhood". Do you think I should remove the `tiered` argument?

Reply:
That's fine with me.